### DISCUSSION TITLE
Should OnClose run before ShutdownEvent?

### DISCUSSION THREAD

**Discussion by krzemienmichal** [micronaut-projects/micronaut-core#8303] [Q&A]
Hello,

I'm using Micronaut version 3.2.6
I have Websocket server application and I've got a problem which I'm not sure if it's a bug or it works as intended.
To be specific inside my Websocket server class I'm defining ```@OnClose``` method and custom onShutdown method which is annotated with ```@EventListener``` and uses ApplicationShutdownEvent. In case of application closure without any open connections custom shutdown methods work fine, however if there's any open connection ```@OnClose``` method runs firstly and close connection with AbnormalClosure reason, then shutdown event works. 
Is this correct behavior? Do you have any ideas how I can solve this in such a way that shutdown event will work before OnClose and send correct ClosureReason?
I found that someone had similar problem in thread #6354 but it's and old thread and it was not answered there.
Let me know if you need more details.

Thanks for help in advance.

**Comment by yawkat** [comment]
How are you stopping the server? You could use your own LifeCycle bean

**Reply by krzemienmichal** [reply]
It doesn't matter if stopped it with ctrl+c, kill process from cli or terminate application from IDE, it works the same way. Ultimately this server should run on k8s pod so I'm trying to handle graceful shutdown for pods restarting.

**Reply by yawkat** [reply]
Then I think your only option is a custom `EmbeddedApplication`, but that's a pain.
